File: other/02/puzzle02.py
input02 = open("input02", encoding="utf-8").read().strip()

instructions = input02.split("\n")

position = 0
depth = 0
for inst in instructions:
    direction, value_str = inst.split(" ")
    value = int(value_str)

    if direction == "forward":
        position += value
    elif direction == "up":
        depth -= value
    elif direction == "down":
        depth += value
    else:
        raise ValueError(f"unexpected direction: {direction}")

print(f"position = {position}, depth = {depth}") # 2003, 980
p1_answer = position * depth
print(f"(p1 answer) position * depth = {p1_answer}") # 1962940


position = 0
depth = 0
aim = 0
for inst in instructions:
    direction, value_str = inst.split(" ")
    value = int(value_str)

    if direction == "down":
        aim += value
    elif direction == "up":
        aim -= value
    elif direction == "forward":
        position += value
        depth += value * aim
    else:
        raise ValueError(f"unexpected direction: {direction}")

print(f"position = {position}, depth = {depth}") # 2003, 905474
p2_answer = position * depth
print(f"(p2 answer) position * depth = {p2_answer}") # 1813664422

File: aoc2021kotlin/src/main/kotlin/Puzzle02.kt
package net.blergh.advent2021

import java.io.File

enum class Direction {
    FORWARD, UP, DOWN
}

data class Instruction(
    val direction: Direction,
    val value: Int
)

object Puzzle02 {
    fun run() {
        val input02 = File("//wsl\$/Debian/home/yettern/apps/advent-of-code-2021/02/input02").readText().trim()
        val instructions = input02.split("\n").map {
            val (directionString, valueString) = it.split(" ")
            Instruction(
                direction=Direction.valueOf(directionString.uppercase()),
                value=Integer.parseInt(valueString)
            )
        }

        part1(instructions)
        part2(instructions)
    }

    private fun part1(instructions: List<Instruction>) {
        var position = 0
        var depth = 0
        for(inst in instructions) {
            when(inst.direction) {
                Direction.FORWARD -> position += inst.value
                Direction.UP -> depth -= inst.value
                Direction.DOWN -> depth += inst.value
            }
        }
        println("position = $position, depth = $depth") // 2003, 980
        val p1Answer = position * depth
        println("(p1 answer) position * depth = $p1Answer") // 1962940
    }

    private fun part2(instructions: List<Instruction>) {
        var aim = 0
        var position = 0
        var depth = 0
        for(inst in instructions) {
            when(inst.direction) {
                Direction.DOWN -> aim += inst.value
                Direction.UP -> aim -= inst.value
                Direction.FORWARD -> {
                    position += inst.value
                    depth += aim * inst.value
                }
            }
        }
        println("position = $position, depth = $depth, aim = $aim") // 2003, 905474, 980
        val p2Answer = position * depth
        println("(p2 answer) position * depth = $p2Answer") // 1813664422
    }
}